content/blog/open-source-monitoring-at-the-hpe-customer-technology-center-böblingen.md

```diff
@@ -0,0 +1,383 @@
+                display_results(node, inode, server_metrics, monitor_urls[server])
+            except Exception as ex:
+                log=logopen(input['lfile'])
+                logwriter(log,'Exception')
+                logwriter(log,str(ex.__context__))
+                logclose(log)
+                pass
+        t1 = time.time()
+        delta.set((t1-t0))
+        while ((t1-t0) < input['mintervall']):
+            time.sleep(1.0)
+            t1 = time.time()   
+        # update once per day the input
+        if ( t1 - start0 > 86400):
+            start0 = t1
+            input = getServerList() 
+            monitor_urls = get_server_urls(input['user'], input['password'], input['server'], input['lfile'])
+```
+
+As you can see, I used four Python functions within the main routine:
+
+* *getServerList*
+* *get\_server\_urls*
+* *get\_server\_data*
+* *display\_results*
+
+The first one - *getServerList* - is just reading the input data out of an XML file. Since this is a pretty standard task in Python, I do not want to discuss this function here further. It should be sufficient to know, that it is reading the list of server IPs and additional runtime parameters (username, password, monitoring intervall, …) into a Python directory that is used in the subsequent parts of the script. I want to focus on the other three functions that are more specific to the ILO Prometheus exporter.
+
+## g﻿et\_server\_urls
+
+The get\_server\_urls routine uses  the list of ILO IP addresses to gather the correct URLs out of the Redfish resource directory. If the ILO is alive  a Redfish client object for this ILO is created and the Redfish resource directory is read. The client object is stored together with the URLs for the thermal, power and computer system information in the python directory server\_urls, which is the return value of the routine.
+
+```python
+def get_server_urls( login_account, login_password, server, lfile):
+
+    server_urls={}
+    if LINUX:
+        cmd='ping -c 2 '
+    log=logopen(lfile)
+    logwriter(log,'get_server_urls')
+    for s in server:
+        ilo_url="https://"+s['ilo']
+        s["url"]=ilo_url
+        if LINUX:
+            response=os.system(cmd+s['ilo'])   # works on Linux without root priviliges
+        else:
+            response = ping(address=s['ilo'],count=1)  # requires root priviliges on Linux
+            if(response.is_alive):
+                response = 0
+            else:
+                response = 256
+        if (response == 0):        
+            try:
+                # Create a Redfish client object
+                REDFISHOBJ = redfish.redfish_client(base_url=ilo_url, username=login_account, password=login_password)  
+                # Login with the Redfish client
+                REDFISHOBJ.login()
+                s["redfish"]=REDFISHOBJ
+                resource_instances = get_resource_directory(REDFISHOBJ, lfile)
```

Comment:
I am surprised to see a call to  the `get_resource_directory` function.  You said above: "...I decided to use the DMTF Redfish Python library..." . However this function is HPE proprietary and not part of the DMTF library.